torch/csrc/jit/passes/freeze_module.cpp

```diff
@@ -0,0 +1,215 @@
+#include <torch/csrc/jit/jit_log.h>
+#include <torch/csrc/jit/passes/freeze_module.h>
+
+#include <torch/csrc/jit/graph_executor_impl.h>
+#include <torch/csrc/jit/passes/inliner.h>
+
+#include <stack>
+
+namespace torch {
+namespace jit {
+
+namespace {
+
+class AttributePropagator {
+ private:
+  // Contains attributes that can't be folded or user directs to keep them.
+  std::map<script::Module*, std::set<std::string>> preservedAttrs_;
+
+  void insertModifiedAttr(script::Module& module, std::string name) {
+    auto it = preservedAttrs_.find(&module);
+    if (it != preservedAttrs_.end()) {
+      it->second.insert(name);
+    } else {
+      std::set<std::string> names;
+      names.insert(name);
+      preservedAttrs_[&module] = names;
+    }
+  }
+
+  // We can use a more efficient algorithm to hash each constant GetAttr to its
+  // conrresponding Value. Based on initial test on resnet50 and other torch
+  // vision tests. GetAttrs are not too frequent so it is ok to chase GetAttr
+  // chain to retrieve their values.
+  bool findConstantAttr(
```

Comment:
The comment above explains some implementation details - it's probably good to start with a general description of what the method does before diving into details. Other functions would benefit from some high-level comment too.